[Sema] `Differentiable` conformance derivation for class types.

dan-zheng opened the pull request:
`Differentiable` derived conformances now supports class types.
Synthesis works just like for struct types.

Class differentiation support requires further differentiation transform changes.

Resolves [TF-630](https://bugs.swift.org/browse/TF-630).

Next steps:
* [TF-631](https://bugs.swift.org/browse/TF-631): Enable differentiation of class methods (handle `ref_element_addr` in differentiation transform).
* [TF-633](https://bugs.swift.org/browse/TF-633): Update `tensorflow-swift-apis` using `Differentiable`-conforming class types.

---

```swift
class Example<T : Differentiable> : Differentiable {
  var x, y: T
  @noDerivative var flag: Bool = true

  // Unlike structs, classes do not get synthesized memberwise initializers.
  // User must define initializer or provide default values for stored properties.
  init(x: T, y: T) {
    self.x = x
    self.y = y
  }

  // Compiler synthesizes (omitting `AllDifferentiableVariables` code):
  //
  // struct TangentVector : Differentiable, AdditiveArithmetic {
  //   var x: T.TangentVector
  //   var y: T.TangentVector
  //   init(x: T.TangentVector, y: T.TangentVector)
  //   typealias TangentVector = Example<T>.TangentVector
  //   final static var zero: Example<T>.TangentVector { get }
  //   static func + (lhs: Example<T>.TangentVector, rhs: Example<T>.TangentVector) -> Example<T>.TangentVector
  //   static func - (lhs: Example<T>.TangentVector, rhs: Example<T>.TangentVector) -> Example<T>.TangentVector
  //   @_implements(Equatable, ==(_:_:)) static func __derived_struct_equals(_ a: Example<T>.TangentVector, _ b: Example<T>.TangentVector) -> Bool
  // }
  //
  // func move(along direction: Example<T>.TangentVector) {
  //   x.move(along: direction.x)
  //   y.move(along: direction.y)
  // }
}
```

dan-zheng committed:
[Sema] `Differentiable` conformance derivation for class types.

`Differentiable` derived conformances now supports class types.
Synthesis works just like for struct types.

Class differentiation support requires further differentiation
transform changes.

Resolves TF-630.

dan-zheng commented on test/Sema/class_differentiable.swift:
```diff
@@ -0,0 +1,570 @@
+// SWIFT_ENABLE_TENSORFLOW
```
These tests are not the best, I directly copied `test/Sema/struct_differentiable.swift`.
Todo: clean up and tighten tests. Fuzzing would be cool.

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei reviewed (commented):
Could you add a test that guarantees that `TangentVector` will never be derived to be `Self` for classes, even if `Self` conforms to `AdditiveArithmetic`?

dan-zheng commented:
> Could you add a test that guarantees that `TangentVector` will never be derived to be `Self` for classes, even if `Self` conforms to `AdditiveArithmetic`?

I actually didn't implement that logic, will do so now.

dan-zheng committed:
Never synthesize `TangentVector = Self` for class types.

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -2869,9 +2869,9 @@ ERROR(compiler_evaluable_ref_non_compiler_evaluable,none,
-ERROR(noderivative_only_on_stored_properties_in_differentiable_structs,none,
-      "'@noDerivative' is only allowed on stored properties in structure types "
-      "that declare a conformance to 'Differentiable'", ())
+ERROR(noderivative_only_on_stored_properties_in_differentiable_types,none,
```
This is a weird name. Consider being explicit about structs and classes. Shortening can be done if you change "stored_properties" to "fields".
```suggestion
ERROR(noderivative_only_on_differentiable_structs_or_class_fields,none,
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/25914/commits/9182edaa6938200b72facd2e7359e084ff808223.

dan-zheng committed:
Minor edits.

Fix validation-test/ParseableInterface/verify_all_overlays.py.
Address review comments.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
